File: app/src/kit/match.ts
/* ONE FORGIVING MATCHER FOR EVERY SEARCH IN THE APP (the owner's ruling, 2026-09-23).
 *
 * The filtering review found four search fields with four rules (FLT-06): Sales matched the
 * literal string in the name only, Graveyard the literal string over five fields, Orders a
 * number the row does not draw, and the store's own search wanted the leading zeros of a
 * collector number (FLT-04). A person types what they read off the card or the row. This
 * answers that one way everywhere a list is filtered in the browser.
 *
 * THE RULES, IN ORDER. They are stated here, not only coded, so that any other search that
 * must agree with this one can be checked against them.
 *
 *  1. FOLD. A string is NFKC-normalized (full-width `５４` is `54`), then NFKD-normalized, its
 *     combining marks are removed, and it is lower-cased. Apostrophes (' ’ ʼ ` ´) are removed,
 *     so `Farfetch'd` is `farfetchd`. Every other run of characters that is not a letter or a
 *     digit becomes one space: `Ho-Oh ex` is `ho oh ex`, `Akali, Deadly` is `akali deadly`.
 *     The COMPACT form of a string is its fold with the spaces removed: `hoohex`.
 *  2. SPLIT. The query is NFKC-normalized, commas become spaces, and it is cut on white space.
 *     Each token loses the punctuation at its ends, except a `#` or a `/` in front.
 *  3. EVERY TOKEN MUST MATCH, IN ANY ORDER. An empty query matches every row. Two tokens next
 *     to each other may match TOGETHER as one card number (rule 4), so `54 132` and `swsh 050`
 *     work. Every way to cover the tokens is tried.
 *  4. A CARD NUMBER is compared on its CANONICAL form: each `/` part is folded, its spaces are
 *     removed, and the zeros in front of its first digit run are removed. `054/132` is
 *     `54/132`, `SWSH050` is `swsh50`, `OP01-001` is `op1001`.
 *      - A token with a `/` matches a card number with the same canonical form.
 *      - A token with no `/` matches the whole number, or its first part: `54` finds `054/132`.
 *      - A hyphen between two digits may also stand for the `/`: `054-132` finds `054/132`.
 *      - A token that starts with `/` matches the second part: `/132` finds every `N/132`.
 *      - Two tokens next to each other match as one number joined by `/` or by nothing.
 *      - Never a substring: `54` does not find `154/200`.
 *  5. A BOX IS FOUND BY ITS NAME ONLY, which is text (rule 7). Its number is never shown and
 *     never searched (the owner's ruling, 2026-09-23): `4` and `B4` do not find a box named
 *     `Mixed Singles` whose number is 4. A box nobody named carries the stored name `Box 4`, so
 *     `box 4` still finds it, by name.
 *  6. A SKU. A token of three or more digits matches a SKU that STARTS with it, zeros kept.
 *  7. TEXT. Otherwise the token is folded (rule 1) into words, and every word must be found in
 *     some text field:
 *      - A word with a letter is found as a substring of a folded field. Mid-word works:
 *        `ventor` finds `Inventor`.
 *      - A word of digits only is found as a WHOLE word or the START of a word, zeros in front
 *        ignored. Typed with no zeros in front, it is the start of a word with its zeros
 *        dropped: `12` finds `00012` and `123`, never `112`. Typed with zeros in front, it is
 *        the start of the word as written, or equal once the zeros go: `0001` finds `00012`, so
 *        an order label narrows as it is typed, and `0002` never finds `26`. A word of only
 *        zeros (`0`, `00`) finds every word that starts with it.
 *     As a fallback, a token with a letter in it matches when its compact form is a substring
 *     of a field's compact form: `hooh` finds `Ho-Oh ex`, `porygonz` finds `Porygon-Z`.
 *     Text fields are what the row draws: a name, a set, a condition (`Damaged`), a buyer, a
 *     note, an order label, a box name.
 *
 * PURE. No React, no state, no network: the same function answers in a spec and on a screen. */

/** The fields of one row a search may match, grouped by how each is compared. */
export type MatchFields = {
  /** What the row draws as words: a name, a set, a condition, a buyer, a label. Rule 7. */
  readonly text?: readonly (string | null | undefined)[]
  /** Collector numbers, as the card prints them (`054/132`). Rule 4. */
  readonly numbers?: readonly (string | null | undefined)[]
  /** TCGplayer SKUs. Rule 6. */
  readonly skus?: readonly (string | number | null | undefined)[]
  /** The boxes the row is in. Only the NAME is matched, as text (rules 5 and 7); `box` is
   *  kept so a caller can hand the record over whole, and is never searched. */
  readonly boxes?: readonly { readonly box: number; readonly name?: string | null }[]
}

const APOSTROPHES = /['’ʼ`´]/gu

/** Rule 1: NFKC, NFKD, no marks, lower case, no apostrophes, other non-alphanumeric runs to
 *  one space. */
export function foldText(value: string): string {
  return value
    .normalize('NFKC')
    .normalize('NFKD')
    .replace(/\p{M}+/gu, '')
    .toLowerCase()
    .replace(APOSTROPHES, '')
    .replace(/[^\p{L}\p{N}]+/gu, ' ')
    .trim()
}

/** Rule 1's compact form: the fold with no spaces. */
export function compactText(value: string): string {
  return foldText(value).replace(/ /gu, '')
}

/** The zeros in front of the first digit run go: `tg05` is `tg5`, `054` is `54`. */
function dropLeadingZeros(part: string): string {
  return part.replace(/^(\p{L}*?)0+(?=\d)/u, '$1')
}

/** Rule 4: a collector number's canonical form. `054/132` and `54/132` are both `54/132`. */
export function canonicalNumber(value: string): string {
  return value
    .normalize('NFKC')
    .split('/')
    .map((part) => dropLeadingZeros(compactText(part)))
    .join('/')
}

/** What a canonical card number may look like: letters, digits, an optional letter, and an
 *  optional second part of the same shape. */
const NUMBER_SHAPE = /^\p{L}{0,6}\d+\p{L}{0,2}(?:\/\p{L}{0,6}\d+\p{L}{0,2})?$/u
const SKU_SHAPE = /^\d{3,}$/u
const DIGITS = /^\d+$/u
const HAS_DIGIT = /\d/u
const HAS_LETTER = /\p{L}/u
/** The punctuation a token loses at its ends. `#` and `/` survive in front (rule 4). */
const EDGE_PUNCTUATION = /^[^\p{L}\p{N}#/]+|[^\p{L}\p{N}]+$/gu

/** Rule 2: the raw tokens of a query. */
export function queryTokens(query: string): string[] {
  /* Commas become spaces first, then one white-space split: a query is not a CSV row. */
  return query
    .normalize('NFKC')
    .replace(/,/gu, ' ')
    .split(/\s+/u)
    .map((token) => token.replace(EDGE_PUNCTUATION, ''))
    .filter((token) => token !== '' && token !== '#' && token !== '/')
}

function present<T>(value: T | null | undefined): value is T {
  return value !== null && value !== undefined
}

type NumberParts = { readonly whole: string; readonly first: string; readonly second: string | null }

/** A row, prepared once for many tokens. */
type Prepared = {
  readonly folded: readonly string[]
  readonly compact: readonly string[]
  /** Every digits-only word of every folded text field, as written and with its zeros in
   *  front dropped, for the digits-only rule. */
  readonly words: readonly { readonly raw: string; readonly bare: string }[]
  readonly numbers: readonly NumberParts[]
  readonly skus: readonly string[]
}

function prepare(fields: MatchFields): Prepared {
  const text = [...(fields.text ?? []).filter(present), ...(fields.boxes ?? []).map((one) => one.name).filter(present)]
  const folded = text.map(foldText).filter((one) => one !== '')
  const words: { raw: string; bare: string }[] = []
  for (const one of folded) {
    for (const word of one.split(' ')) if (DIGITS.test(word)) words.push({ raw: word, bare: dropLeadingZeros(word) })
  }
  return {
    folded,
    compact: folded.map((one) => one.replace(/ /gu, '')),
    words,
    numbers: (fields.numbers ?? [])
      .filter(present)
      .map(canonicalNumber)
      .filter((one) => one !== '')
      .map((whole) => {
        const [first = '', second] = whole.split('/')
        return { whole, first, second: second ?? null }
      }),
    skus: (fields.skus ?? []).filter(present).map((sku) => String(sku).trim()),
  }
}

/** Rule 4, for one raw token (a `#` or `/` in front allowed). */
function numberMatch(raw: string, row: Prepared): boolean {
  if (row.numbers.length === 0) return false
  if (raw.startsWith('/')) {
    const second = canonicalNumber(raw.slice(1))
    if (!HAS_DIGIT.test(second) || second.includes('/')) return false
    return row.numbers.some((number) => number.second === second)
  }
  const bare = raw.replace(/^#/u, '')
  if (!HAS_DIGIT.test(bare)) return false
  const forms = new Set([canonicalNumber(bare), canonicalNumber(bare.replace(/(\d)-(?=\d)/gu, '$1/'))])
  for (const form of forms) {
    if (!NUMBER_SHAPE.test(form)) continue
    const whole = form.includes('/')
    if (row.numbers.some((number) => number.whole === form || (!whole && number.first === form))) return true
  }
  return false
}

/** Rule 7's digits-only word: the start of a digits-only text word, as written or with the zeros
 *  in front of both dropped. */
function digitWordMatch(word: string, row: Prepared): boolean {
  /* Only zeros so far (`0`, `00`): the first keys of `09-03-26_…` or `00012`. Any word that
   * starts with them matches, so the first key never empties the list. */
  if (/^0+$/u.test(word)) return row.words.some((one) => one.raw.startsWith(word))
  const bare = dropLeadingZeros(word)
  /* A word typed WITH zeros in front is being typed as written, so it is the start of the word
   * as written (`0001` of `00012`), or equal once the zeros are dropped. A word typed without
   * them is the start of a word whose zeros are dropped (`12` of `00012`). So `0002` never
   * finds `26`, and `12` never finds `112`. */
  if (word !== bare) return row.words.some((one) => one.raw.startsWith(word) || one.bare === bare)
  return row.words.some((one) => one.bare.startsWith(bare))
}

/** Rule 7, for one raw token. */
function textMatch(raw: string, row: Prepared): boolean {
  const words = foldText(raw).split(' ').filter((word) => word !== '')
  if (words.length === 0) return false
  const each = words.every((word) =>
    DIGITS.test(word) ? digitWordMatch(word, row) : row.folded.some((field) => field.includes(word)),
  )
  if (each) return true
  const compact = words.join('')
  return HAS_LETTER.test(compact) && row.compact.some((field) => field.includes(compact))
}

/** Rules 4 to 7 for one token alone. */
function tokenMatch(raw: string, row: Prepared): boolean {
  if (SKU_SHAPE.test(raw) && row.skus.some((sku) => sku.startsWith(raw))) return true
  if (numberMatch(raw, row)) return true
  if (raw.startsWith('/')) return false
  return textMatch(raw, row)
}

/** Rule 3: two neighbouring tokens that match together as one card number. */
function pairMatch(left: string, right: string, row: Prepared): boolean {
  if (left.startsWith('/') || right.startsWith('/') || right.startsWith('#')) return false
  return numberMatch(`${left}/${right}`, row) || numberMatch(`${left}${right}`, row)
}

/** Can the tokens from `at` on all be matched, singly or in neighbouring pairs? */
function cover(tokens: readonly string[], at: number, row: Prepared): boolean {
  if (at >= tokens.length) return true
  const one = tokens[at] as string
  if (tokenMatch(one, row) && cover(tokens, at + 1, row)) return true
  const two = tokens[at + 1]
  return two !== undefined && pairMatch(one, two, row) && cover(tokens, at + 2, row)
}

/** Does this row match this query? Every token must match, in any order (rules 2 to 7). */
export function matchQuery(query: string, fields: MatchFields): boolean {
  const tokens = queryTokens(query)
  if (tokens.length === 0) return true
  return cover(tokens, 0, prepare(fields))
}

/** The rows that match, in their own order. `fieldsOf` says what each row draws. */
export function filterByQuery<T>(rows: readonly T[], query: string, fieldsOf: (row: T) => MatchFields): T[] {
  const tokens = queryTokens(query)
  if (tokens.length === 0) return [...rows]
  return rows.filter((row) => cover(tokens, 0, prepare(fieldsOf(row))))
}

File: server/match.py
"""One forgiving matcher, ported from `app/src/kit/match.ts` (the owner's ruling, 2026-09-23,
FLT-06/04): name words in any order, card numbers with or without leading zeros, a hyphen
standing for the slash, accents folded, punctuation folded, SKUs and box names.

THE RULES ARE STATED IN `kit/match.ts`'S OWN DOCSTRING AND ARE NOT RESTATED HERE. This module
is a same-behaviour port, not a second design — `scripts/match-selftest.py` proves the two
agree by running every row of `app/src/kit/match.cases.json` (the case table the filtering
lane writes) through both. Where the two disagree, this file is wrong; fix it here, never the
case table.

STDLIB ONLY. `unicodedata` gives NFKC/NFKD and combining-mark removal without the `regex`
package's `\\p{L}`/`\\p{N}` classes — `requirements.txt`'s own "Deliberately absent" section is
the standing reason no dependency is added for this. `str.isalpha()` is Unicode-aware in
Python and stands in for `\\p{L}` closely enough, matching `kit/match.ts`'s own `\\p{L}` (its
`NUMBER_SHAPE`/etc. carry the `u` flag, which widens `\\p{L}` to every Unicode letter). DIGITS
ARE THE ONE PLACE THAT SUBSTITUTION BREAKS (S4, the Opus review, 2026-09-25): `str.isdigit()`
and Python's bare `\\d` are BOTH Unicode-aware, but JS's `\\d` is not — it is always `[0-9]`,
`u` flag or not, because JS's Unicode regex mode widens `\\p{...}` property escapes and does
nothing to `\\d` itself. So every digit test in this file (`_has_digit`, `_DIGITS_ONLY`,
`_SKU_SHAPE`, `_HYPHEN_BETWEEN_DIGITS`, and the two digit runs inside `_drop_leading_zeros`/
`_number_shape_ok`) is deliberately ASCII-only (`_ASCII_DIGITS`, or a `[0-9]` character class
rather than `\\d`), even though nothing else here is. `str.isalnum()` (used only inside
`fold_text`, never for a digit-specific decision) stands in for `\\p{L}`/`\\p{N}` combined, and
is left Unicode-aware on purpose — it is what a query keeps rather than what it treats as a
number, so the same asymmetry does not apply to it.

A KNOWN, UNFIXED GAP THE ASCII FIX DOES NOT TOUCH: `kit/match.ts:foldText` deletes every
Unicode MARK outright (`.replace(/\\p{M}+/gu, '')`), whatever its combining class. This
file's own mark strip (`unicodedata.combining(ch) != 0`, inside `fold_text`) removes only a
mark with a NON-ZERO canonical combining class — a genuine COMBINING mark, the kind NFKD
produces for an accent. A mark whose class IS zero ("ccc-0", a SPACING mark, e.g. Devanagari's
own vowel signs) is not a combining mark by that test, so it survives to the next step
instead — and since a Mark character is never `str.isalnum()`, that step folds it into a
WORD BREAK (one space) rather than deleting it, splitting a word the JS side would have kept
joined. No case in `app/src/kit/match.cases.json` has ever needed a ccc-0 mark folded, so
this is recorded rather than chased: `docs/decisions/D271-one-forgiving-search-matcher.md`
carries it.

USED TWO WAYS. `scripts/match-selftest.py` calls `match_query` directly against the generic
`MatchFields` shape (text/numbers/skus/boxes), the same contract `kit/match.ts` takes.
`server/capture_server.py:_match_rank` calls the lower-level `fold_text`/`canonical_number`
primitives on the SIX fields it has always ranked (name, number, sku, set_hint, note, the
composed key), because that function keeps its own three-tier rank (exact number, name
prefix, substring) rather than a plain boolean — `do_search`'s sort order depends on the
three tiers staying distinct, which a single `match_query() -> bool` cannot express.

PURE. No I/O, no store, no network — the same function answers in a script and in the server.
"""

from __future__ import annotations

import re
import unicodedata
from typing import Dict, Iterable, List, Optional, Sequence, TypedDict, Union

Number = Union[str, int, float, None]


class MatchFields(TypedDict, total=False):
    """One row a search may match, grouped by how each field is compared — the same shape
    `app/src/kit/match.ts`'s own `MatchFields` type declares."""

    text: Sequence[Optional[str]]
    numbers: Sequence[Optional[str]]
    skus: Sequence[Number]
    boxes: Sequence[Dict[str, object]]


_APOSTROPHES = re.compile(r"['’ʼ`´]")
# ASCII DIGITS ONLY (S4, the Opus review, 2026-09-25), MATCHING `\d` ON THE OTHER SIDE.
# `app/src/kit/match.ts`'s own `\d` is JS regex syntax, which is ALWAYS `[0-9]` — the `u`
# flag on `NUMBER_SHAPE`/`SKU_SHAPE`/`DIGITS`/`HAS_DIGIT` widens `\p{L}` to every Unicode
# letter, but it does nothing to `\d`, which JS has no Unicode-digit mode for at all.
# Python's bare `\d` (no `re.ASCII`) and `str.isdigit()` are both Unicode-aware — `"५"`
# (Devanagari digit five, U+096B) is `True` on both and NFKC leaves it untouched (it is not
# a compatibility decomposition of an ASCII digit, unlike a full-width `５`, which NFKC
# already folds before either side ever sees it) — so a query built from a real Unicode
# digit would have canonicalized as a number on the SERVER and stayed inert TEXT in the
# BROWSER: the one place this repo's own two-implementation contract could silently break.
_ASCII_DIGITS = "0123456789"
_DIGITS_ONLY = re.compile(r"^[0-9]+$")
_ZEROS_ONLY = re.compile(r"^0+$")
_SKU_SHAPE = re.compile(r"^[0-9]{3,}$")
_HYPHEN_BETWEEN_DIGITS = re.compile(r"([0-9])-(?=[0-9])")


def fold_text(value: str) -> str:
    """Rule 1: NFKC, NFKD, no combining marks, lower case, no apostrophes, every other run of
    non-alphanumeric characters becomes one space, trimmed."""
    value = unicodedata.normalize("NFKC", str(value))
    value = unicodedata.normalize("NFKD", value)
    value = "".join(ch for ch in value if not unicodedata.combining(ch))
    value = value.lower()
    value = _APOSTROPHES.sub("", value)
    out: List[str] = []
    spaced = False
    for ch in value:
        if ch.isalnum():
            out.append(ch)
            spaced = False
        elif not spaced:
            out.append(" ")
            spaced = True
    return "".join(out).strip()


def compact_text(value: str) -> str:
    """Rule 1's compact form: the fold with no spaces."""
    return fold_text(value).replace(" ", "")


def _has_digit(value: str) -> bool:
    return any(ch in _ASCII_DIGITS for ch in value)


def _has_letter(value: str) -> bool:
    return any(ch.isalpha() for ch in value)


def _drop_leading_zeros(part: str) -> str:
    """The zeros in front of the first digit run go, stopping at the last digit: `054` is
    `54`, `tg05` is `tg5`, and a run of zeros alone (`000`) keeps its last one (`0`), because
    that last zero has no digit after it to license stripping it too — the same lookahead
    `kit/match.ts:dropLeadingZeros` reads off its own regex."""
    i = 0
    n = len(part)
    while i < n and part[i].isalpha():
        i += 1
    letters = part[:i]
    j = i
    while j < n - 1 and part[j] == "0" and part[j + 1] in _ASCII_DIGITS:
        j += 1
    return letters + part[j:]


def _hyphen_to_slash(value: str) -> str:
    """A hyphen directly between two digits may stand for the collector-number slash:
    `054-132` reads the same as `054/132`."""
    return _HYPHEN_BETWEEN_DIGITS.sub(r"\1/", value)


def canonical_number(value: str) -> str:
    """Rule 4: a collector number's canonical form. `054/132` and `54/132` are both `54/132`."""
    value = unicodedata.normalize("NFKC", str(value))
    return "/".join(_drop_leading_zeros(compact_text(part)) for part in value.split("/"))


def _number_shape_ok(part: str) -> bool:
    """One side of `NUMBER_SHAPE`: 0-6 leading letters, one or more digits, 0-2 trailing
    letters — nothing else."""
    i = 0
    n = len(part)
    letters = 0
    while i < n and part[i].isalpha() and letters < 6:
        i += 1
        letters += 1
    if i < n and part[i].isalpha():
        return False
    digits_start = i
    while i < n and part[i] in _ASCII_DIGITS:
        i += 1
    if i == digits_start:
        return False
    trailing = 0
    while i < n and part[i].isalpha() and trailing < 2:
        i += 1
        trailing += 1
    if i < n and part[i].isalpha():
        return False
    return i == n


def _is_number_shape(value: str) -> bool:
    parts = value.split("/")
    if len(parts) > 2:
        return False
    return all(_number_shape_ok(part) for part in parts)


class _NumberParts(TypedDict):
    whole: str
    first: str
    second: Optional[str]


class _Prepared(TypedDict):
    folded: List[str]
    compact: List[str]
    words: List[Dict[str, str]]
    numbers: List[_NumberParts]
    skus: List[str]


def _number_parts(canon: str) -> _NumberParts:
    if "/" in canon:
        first, second = canon.split("/", 1)
        return {"whole": canon, "first": first, "second": second}
    return {"whole": canon, "first": canon, "second": None}


def _present(values: Optional[Iterable]) -> List:
    return [v for v in (values or []) if v is not None and v != ""]


def _prepare(fields: MatchFields) -> _Prepared:
    box_names = [b.get("name") for b in (fields.get("boxes") or []) if b.get("name")]
    text_values = [str(t) for t in _present(fields.get("text"))] + [str(n) for n in box_names]
    folded = [f for f in (fold_text(t) for t in text_values) if f]
    compact = [f.replace(" ", "") for f in folded]
    words: List[Dict[str, str]] = []
    for one in folded:
        for word in one.split(" "):
            if word and _DIGITS_ONLY.match(word):
                words.append({"raw": word, "bare": _drop_leading_zeros(word)})
    numbers: List[_NumberParts] = []
    for raw_number in _present(fields.get("numbers")):
        canon = canonical_number(str(raw_number))
        if canon:
            numbers.append(_number_parts(canon))
    skus = [str(sku).strip() for sku in _present(fields.get("skus"))]
    return {"folded": folded, "compact": compact, "words": words, "numbers": numbers, "skus": skus}


def _number_match(raw: str, numbers: Sequence[_NumberParts]) -> bool:
    """Rule 4, for one raw token (a leading `#` or `/` allowed)."""
    if not numbers:
        return False
    if raw.startswith("/"):
        second = canonical_number(raw[1:])
        if not _has_digit(second) or "/" in second:
            return False
        return any(n["second"] == second for n in numbers)
    bare = raw[1:] if raw.startswith("#") else raw
    if not _has_digit(bare):
        return False
    forms = {canonical_number(bare), canonical_number(_hyphen_to_slash(bare))}
    for form in forms:
        if not _is_number_shape(form):
            continue
        whole = "/" in form
        if any(n["whole"] == form or (not whole and n["first"] == form) for n in numbers):
            return True
    return False


def _sku_match(raw: str, skus: Sequence[str]) -> bool:
    return bool(_SKU_SHAPE.match(raw)) and any(sku.startswith(raw) for sku in skus)


def _digit_word_match(word: str, words: Sequence[Dict[str, str]]) -> bool:
    """Rule 7's digits-only word: the start of a digits-only text word, as written or with
    the zeros in front of both dropped."""
    if _ZEROS_ONLY.match(word):
        return any(w["raw"].startswith(word) for w in words)
    bare = _drop_leading_zeros(word)
    if word != bare:
        return any(w["raw"].startswith(word) or w["bare"] == bare for w in words)
    return any(w["bare"].startswith(bare) for w in words)


def _text_match(raw: str, row: _Prepared) -> bool:
    words = [w for w in fold_text(raw).split(" ") if w]
    if not words:
        return False

    def word_ok(word: str) -> bool:
        if _DIGITS_ONLY.match(word):
            return _digit_word_match(word, row["words"])
        return any(word in field for field in row["folded"])

    if all(word_ok(w) for w in words):
        return True
    compact = "".join(words)
    return _has_letter(compact) and any(compact in field for field in row["compact"])


def _token_match(raw: str, row: _Prepared) -> bool:
    """Rules 4 to 7 for one token alone."""
    if _sku_match(raw, row["skus"]):
        return True
    if _number_match(raw, row["numbers"]):
        return True
    if raw.startswith("/"):
        return False
    return _text_match(raw, row)


def _pair_match(left: str, right: str, row: _Prepared) -> bool:
    """Rule 3: two neighbouring tokens that match together as one card number."""
    if left.startswith("/") or right.startswith("/") or right.startswith("#"):
        return False
    return _number_match(f"{left}/{right}", row["numbers"]) or _number_match(
        f"{left}{right}", row["numbers"]
    )


def _cover(tokens: Sequence[str], row: _Prepared) -> bool:
    """Rule 3: can every token be covered, in order, by the single-token rule or the
    pair rule?

    MEMOIZED ON `at` (F1, the Opus review, 2026-09-25). The un-memoized version tries
    BOTH the single-token branch and the pair branch at every position, and each branch
    recurses into the rest of the tokens — the same sub-problem (`at+1` from the single
    branch consuming one token, and again from wherever the pair branch's own single-try
    lands two tokens later) gets solved again from scratch every time a caller reaches it
    by a different path, which is exactly what makes Fibonacci's own recursion
    exponential. Measured: `("132 " * 27 + "/132 /999")` against a card whose number is
    `132/132` took 3.9s at 26 terms, uncached. `at` alone decides the rest of the answer
    — `tokens` and `row` are fixed for the one `match_query` call this closes over — so a
    dict keyed on `at` turns the call tree back into a line: at most `len(tokens) + 1`
    distinct calls, never two of the same `at` doing the work twice.
    """
    memo: Dict[int, bool] = {}

    def cover(at: int) -> bool:
        if at in memo:
            return memo[at]
        if at >= len(tokens):
            memo[at] = True
            return True
        one = tokens[at]
        result = False
        if _token_match(one, row) and cover(at + 1):
            result = True
        if not result and at + 1 < len(tokens):
            two = tokens[at + 1]
            if _pair_match(one, two, row) and cover(at + 2):
                result = True
        memo[at] = result
        return result

    return cover(0)


_EDGE_PUNCT_KEEP = "#/"


def _strip_edge_punctuation(token: str) -> str:
    """The punctuation a token loses at its ends. `#` and `/` survive in front (rule 4)."""
    start = 0
    n = len(token)
    while start < n and not (token[start].isalnum() or token[start] in _EDGE_PUNCT_KEEP):
        start += 1
    end = n
    while end > start and not token[end - 1].isalnum():
        end -= 1
    return token[start:end]


def query_tokens(query: str) -> List[str]:
    """Rule 2: the raw tokens of a query."""
    query = unicodedata.normalize("NFKC", str(query)).replace(",", " ")
    raw_tokens = [t for t in re.split(r"\s+", query) if t]
    tokens = [_strip_edge_punctuation(t) for t in raw_tokens]
    return [t for t in tokens if t not in ("", "#", "/")]


def match_query(query: str, fields: MatchFields) -> bool:
    """Does this row match this query? Every token must match, in any order (rules 2 to 7)."""
    tokens = query_tokens(query)
    if not tokens:
        return True
    return _cover(tokens, _prepare(fields))
